New Permissions Pages

alxnddr opened the pull request:
## Description

This PR replaces the old Permissions UI with the new one that should make viewing/editing permissions easier especially for customers with a lot of data. It should behave the same in terms of business logic, only the UI has changed.

## Screenshots

### Collections permissions

**Before**
<img width="500" alt="Screen Shot 2021-08-12 at 19 06 45" src="https://user-images.githubusercontent.com/14301985/129230754-260004e7-ebe8-49e3-921d-143c546e8f96.png">

**After**
<img width="500" alt="Screen Shot 2021-08-12 at 01 32 49" src="https://user-images.githubusercontent.com/14301985/129112297-47ffa523-8c68-44f1-8c89-78040009ce33.png">

### Data permissions
**Before**
<img width="500" alt="Screen Shot 2021-08-12 at 19 06 57" src="https://user-images.githubusercontent.com/14301985/129230814-d88e90de-2a5a-43a1-b670-19e08d3d8ade.png">


**After — groups focus** 
<img width="500" alt="Screen Shot 2021-08-09 at 04 04 15" src="https://user-images.githubusercontent.com/14301985/128651604-7ac2c44f-285f-4b05-9d2f-79d788e49330.png">

**After — databases focus** 
<img width="500" alt="Screen Shot 2021-08-09 at 04 04 29" src="https://user-images.githubusercontent.com/14301985/128651612-fe9db2ea-0f67-4d41-9db3-bd646cdd196e.png">

### Collection permissions modal

**Before**
<img width="500" alt="Screen Shot 2021-08-12 at 19 07 48" src="https://user-images.githubusercontent.com/14301985/129230672-27882d82-4601-403f-9b56-d7748e598df5.png">

**After**
<img width="500" alt="Screen Shot 2021-08-09 at 04 05 02" src="https://user-images.githubusercontent.com/14301985/128651620-68f00cbd-a2e1-44b1-91ef-0ca9f6d8ea82.png">

### Snippet folder permissions modal

**Before**
<img width="500" alt="Screen Shot 2021-08-12 at 19 08 06" src="https://user-images.githubusercontent.com/14301985/129230635-9a477d24-96d1-497b-91e2-0b3fd2b524b1.png">

**After**
<img width="500" alt="Screen Shot 2021-08-09 at 04 05 16" src="https://user-images.githubusercontent.com/14301985/128651631-5559a88e-cc07-4792-a116-5109c8b11737.png">

### Warnings

<img width="400" alt="Screen Shot 2021-08-16 at 11 07 25" src="https://user-images.githubusercontent.com/14301985/129532084-ffea4ce0-3bf1-4d3e-b321-5cf2a4e5b0e5.png">

Closes https://github.com/metabase/metabase/issues/4186 
Closes https://github.com/metabase/metabase/issues/8165  
Closes https://github.com/metabase/metabase/issues/10874  
Closes https://github.com/metabase/metabase/issues/10920 
Closes https://github.com/metabase/metabase/issues/12489 
Closes https://github.com/metabase/metabase/issues/14669 
Closes https://github.com/metabase/metabase/issues/14671 
Closes https://github.com/metabase/metabase/issues/17185 
Closes https://github.com/metabase/metabase/issues/15233

alxnddr commented:
@metabase-bot run visual tests

alxnddr commented:
@metabase-bot run visual tests

kdoh commented:
Overall looking quite solid. There are a couple things I noticed while playing with this that I'd consider tweaking.

1. I think we need a more obvious sense that the database names in the top level group based view are clickable. Think it could be worth using `text-decoration: underline` here just to be super obvious that you can drill down further.
<img width="908" alt="Screen Shot 2021-08-13 at 9 46 53 AM" src="https://user-images.githubusercontent.com/5248953/129366719-8bb7e302-7b58-461f-bf80-527335db56f6.png">

2. It's not clear which database I'm looking at when I've drilled through into a DB or schema in the group view. We might want to consider adding breadcrumbs here above the "Schema name" label just to help make it easier to scoot around. This is especially true when you get linked to that screen when selecting which tables you want to limit for a particular group as I mention in the next point.
<img width="1491" alt="Screen Shot 2021-08-13 at 9 53 43 AM" src="https://user-images.githubusercontent.com/5248953/129367686-0280dc5e-cb84-4ecc-ab11-f656eb798f17.png">

3. I've noticed some jumpiness in the top of the permission page when I select "Limit access" and then I get taken over to the group view. It's like the confirm save bar pops in for a second and then disappears. I tried capturing it in this recording. 

https://user-images.githubusercontent.com/5248953/129368895-ed138205-8f87-4d6e-88e4-1ba078b46ce6.mov

alxnddr commented:
@kdoh I've addressed all the points, thanks!

alxnddr committed:
add PermissionSelect component

alxnddr committed:
add PermissionSidebar component

alxnddr committed:
add PermissionTable component

alxnddr committed:
add PermissionEditor component

alxnddr committed:
add useLeaveConfirmation hook

alxnddr committed:
add selectors for the collections permissions page

alxnddr committed:
add selectors for the data permissions pages

alxnddr committed:
add confirmation creators to use in selectors

alxnddr committed:
sandboxing permissions modal: redirect to parent route when closed

alxnddr committed:
Add PermissionsPageLayout component

alxnddr committed:
Adjust CollectionPermissionsModal to use new permissions UI

alxnddr committed:
Add CollectionPermissionsPage

alxnddr committed:
Add data permissions pages

alxnddr committed:
Adjust collection permissions modal route

alxnddr committed:
Adjust permissions routes according new pages structure

alxnddr committed:
update permissions actions, reducers

alxnddr committed:
adjust sandboxing

alxnddr committed:
remove old permissions code

alxnddr committed:
support admin colorScheme in Radio

alxnddr committed:
support admin colorScheme in TextInput

alxnddr committed:
support admin colorScheme in Tree

alxnddr committed:
update permissions specs

alxnddr commented:
@metabase/frontend-developers it is a big one but I split it into granular commits to make the review a bit easier. Please tell if you'd prefer me to create multiple PRs

flamber reviewed (commented):
Just did a quick test of `e44177c` - here's what I've found:
Looks really nice, but takes a while to get used to it.

1. Clicking the database/schema/table links fails - they are `/admin/permissions/data/database1` instead of `/admin/permissions/data/database/1`
![image](https://user-images.githubusercontent.com/1447303/129572469-f4eb2567-e999-46ab-be25-8a1119a4c88e.png)
2. The table search also search schema, but it's a bit strange, when there's overlaps
![image](https://user-images.githubusercontent.com/1447303/129573027-9587c178-7d31-4308-8a21-f568c703be61.png)
![image](https://user-images.githubusercontent.com/1447303/129573092-b6912bb2-e8dd-4087-8af8-60b489edfe40.png)
3. I'm still seeing #15233
4. If I redo #14669, then I can get to a state, where the table is sandboxed, without me defining anything in the sandbox (clicked cancel) and then saved permissions, but now it's not possible to edit the sandbox settings - nothing opens. This might have something to do with sandbox modal sends a GTAP request, which is separate from the overall permission Save request, so it's possible to get into a weird state (though this has always existed and doesn't expose any data incorrectly).

alxnddr commented:
@flamber thanks for catching these, I will fix

alxnddr committed:
fix data permissions url typo

nemanjaglumac commented:
> it is a big one but I split it into granular commits to make the review a bit easier. Please tell if you'd prefer me to create multiple PRs

It's not just about the review. It's much easier to revert and track/bisect changes later on if they are divided in a lot of smaller PRs.

There should **always** be an (epic) issue with a clear description of the whole feature and the checkboxes for the planned implementation. Each PR should correspond to the individual task.

This gives us the ability to track progress much easier.
I vote for splitting this in separate logical PRs, with clear explanation of each of the steps tracked in an "aggregate" issue.

rlotun commented:
> It's not just about the review. It's much easier to revert and track/bisect changes later on if they are divided in a lot of smaller PRs.

Just a reminder that `Rebase and Merge` is turned on as an option for everyone. It should be used sparingly, but in the case of a PR with logical commits it's a good fit - you can get the bisect ability on merge since all the granular commits will appear in master.

A series of smaller PRs might help with review, but there is also a cost to producing them after the fact (overhead for the engineer, CI test runs, etc) - so if this branch is already formed as a series of logical commits, it might be ok, assuming the review isn't too hard. All I ask is that we balance properly between three dimensions: i) try to get at logical well formed commits that support bisecting ii) have a process that's easy to review iii) make it so the software developer isn't going through a lot of overhead and pain to get their work in.

gusaiani commented:
I also think that 

* if the PR is largeish but has well-structured commits
* the PR is still easy to review
* we merge them without squashing

we should be good to go.

Compare that with cases where we would have half-finished features in master while we wait for review of PR 3 of 5 (all tiny) of a feature.

alxnddr committed:
fix sandboxing

alxnddr committed:
add search debounce to permission pages

alxnddr committed:
exclude schemas from search, make it flat

alxnddr commented:
@flamber I addressed your comments, thanks. Speaking of the second point, I exclude schemas from the search because I assume that schemas count is not so significant in comparison to tables count. What do you think about that?

alxnddr commented:
@nemanjaglumac @rlotun @gusaiani 

I totally agree that reviewing such a bug PR might be a tedious process that requires preserving at least 2-3 hours straight for the review, so I was not sure about submitting one bug PR to review.

PROS:
- There is no middle state between old and new permissions. The minimum deliverable thing is to reproduce the functionality. In case of reverting, I would not expect to revert just a part of this PR, and I'd prefer to squash & merge it then.
- In a single PR with the entire functionality, reviewers have the full context. When they want to check how some specific case works or what leads to some UI component state.
- If I created a sequence of PRs right when I have just implemented some piece, it would lead to reviewing unfinished functionality, and the code might change significantly. During the implementation, I refactored it to reduce duplication, etc. So now, overall review time is likely to be lower.

CONS:
- For features that can be delivered to users in multiple stages creating huge PR delays merging independent product increment
- Bisecting
- Harder to track progress, as Nemanja mentioned
- It is just really tough to review huge PRs no matter what are benefits to have it big 

At this point, I already have everything here, but I can just cherry-pick these commits into separate PRs. Will it be beneficial?
Anyway, +3.6k, -3.2k is a lot for review, and I won't be doing that anymore.

flamber reviewed (commented):
@alxnddr Great with the fixes - it definitely makes it a lot easier to play with now ;-)

- It shows a blank page for database that doesn't have schemas (like MySQL), when clicking table in the sidebar - historic issue 12269/12372
`/admin/permissions/data/database/2/schema/null/table/7`

I'm unsure about the search. We have many customers with hundreds of schemas, and many with similar named tables in different schemas, so searching for a table will not show which schema the table belongs to.
But I don't think it should hold back this PR, since it might be easier to make slight adjustments later, when it's merged and tested on many instances (like horrorshow etc).

alxnddr committed:
fix permissions for databases without schemas

alxnddr committed:
fix copy for native queries permissions, show table display_name instead of raw name

alxnddr committed:
fix before leave warning shown when change collections

alxnddr committed:
permissions sidebar: make database name clickable

alxnddr committed:
fix string groupIds lead to bugs in permissions editor

alxnddr committed:
fix disabling table permissions shows all tables permission disabled

alxnddr committed:
fix permissions copy

alxnddr committed:
fix permissions disabled tooltips copy

rlotun commented:
> At this point, I already have everything here, but I can just cherry-pick these commits into separate PRs. Will it be beneficial?

If, as you say, there isn't a good intermediate stage between old and new permissions I'm fine with squash merging it. It'll be a large commit but it will be a logical one at least.

daltojohnso commented on frontend/src/metabase/hoc/ModalRoute.jsx:
```diff
@@ -5,6 +5,15 @@ import { push } from "react-router-redux";
 import { connect } from "react-redux";
 import Modal from "metabase/components/Modal";
 
+export const getParentPath = (route, location) => {
```
can you add a unit test for this function?

daltojohnso commented on frontend/src/metabase/admin/permissions/constants/data-permissions.js:
```diff
@@ -0,0 +1,29 @@
+    label: t`Allowed`,
+    value: "all",
+    icon: "check",
+    iconColor: color("success"),
```
not totally sure if this will work with whitelabeling clobbering what color these strings map to

daltojohnso commented on frontend/src/metabase/admin/permissions/constants/collections-permissions.js:
```diff
@@ -0,0 +1,23 @@
+    label: t`Curate`,
+    value: "write",
+    icon: "check",
+    iconColor: color("success"),
```
mentioned in another comment, but I think whitelabeling might now work with this `color` calls that run on instantiation

daltojohnso commented on frontend/src/metabase/admin/permissions/components/PermissionsEditor/PermissionsEditor.jsx:
```diff
@@ -0,0 +1,103 @@
+}) {
+  const [filter, setFilter] = useState("");
+
+  const handleFilterChange = text => setFilter(text);
```
this function looks redundant to me. you should be able to directly pass `setFilter`

daltojohnso reviewed (changes requested):
nice work! this is a big boy. let me know if any of my comments don't make sense. i'll take another pass through this code later today as well in case i missed anything

it'd also be nice to see unit tests on code like the selectors in this commit https://github.com/metabase/metabase/pull/17328/commits/41acc0e2b59e225d4fa0987d2d1a6c4b094aaf8d and https://github.com/metabase/metabase/pull/17328/commits/815fc1f65e2e4284a4d1cca5213435f5a08fb594 if possible (unless i missed this unit tests somehow)

alxnddr committed:
fix, add specs

alxnddr committed:
fix permissions edit bar button colors, fix breadcrumbs on schemaless dbs

alxnddr committed:
fix official collection color, make sidebar tree scrollable

alxnddr committed:
fix scroll in permissions modal

alxnddr committed:
fix breadcrumbs

alxnddr committed:
change table icon

alxnddr committed:
create helpers for entity ids and permissions url creation

alxnddr committed:
remove hardcoded colors

alxnddr committed:
make confirmations an array

alxnddr committed:
store confirmAction in state

kulyk commented:
Tremendous work @alxnddr 👏
Have some general UX feedback here (cc @kdoh)

**Discard changes popup**

The popup is a little bit confusing. If I have unsaved changes and try to leave a page or change a tab, I see a popup saying "your changes will be discarded if you leave" with two options: "Cancel" and "Discard changes". I expected "Cancel" to close the popup, so I can save my changes and "Discard changes" to leave the page without saving the changes, but it works vice-versa. I guess it doesn't align with the common UX pattern for confirmations like these

<details>
<summary>Video Demo</summary>

https://user-images.githubusercontent.com/17258145/130278046-886d1f80-a085-44b3-8bb6-b745aa00ed61.mov
</details>

**Updating permissions for child collections**

When I change permission for a collection, the popover has a toggle to apply the changes to child collections too. The toggle is placed below the permission options, so I tend to select the permission first and then select the toggle. However:

1. The popover closes once I select the permission level, so I need to open it for the second time to change the toggle
2. If I actually reopen the popover and click the toggle, it won't apply the change to children. It only works if I *first* change the toggle and then change the permission level. I'm just worried it might confuse people

<details>
<summary>Video Demo</summary>

https://user-images.githubusercontent.com/17258145/130278801-8fec2769-8370-447f-8d2e-b991035835e7.mov

</details>

alxnddr committed:
add ModalRoute specs

alxnddr committed:
fix propTypes, typo

alxnddr committed:
build permission pages urls with helpers

alxnddr committed:
revert switching from ref to state for confirm action function due to react invokes it on a render

alxnddr committed:
move mapStateToProps to the top of the files

alxnddr committed:
add visual specs

alxnddr commented:
@metabase-bot run visual tests

alxnddr committed:
add more specs

alxnddr commented on enterprise/frontend/src/metabase-enterprise/sandboxes/index.js:
```diff
@@ -12,43 +13,50 @@ import { push } from "react-router-redux";
+    ? `/admin/permissions/data/group/${groupId}/database/${databaseId}/schema/${encodeURIComponent(
+        schemaName,
+      )}/${tableId}/segmented`
+    : `/admin/permissions/data/group/${groupId}/database/${databaseId}/${tableId}/segmented`;
```
I created helpers which made it a bit more readable

alxnddr commented on frontend/src/metabase/admin/permissions/selectors/collection-permissions.js:
```diff
@@ -0,0 +1,259 @@
+    ? ROOT_COLLECTION.id
+    : parseInt(props.params.collectionId);
+
+const getRootCollectionTreeItem = () => {
+  const rootCollectionIcon = getCollectionIcon(ROOT_COLLECTION);
+  return {
+    ...ROOT_COLLECTION,
+    icon: rootCollectionIcon.name,
+    iconColor: rootCollectionIcon.color,
+  };
+};
```
This is the root item, other folders that came from the BE use icon tooltips

alxnddr commented:
@flamber @daltojohnso @kulyk thank you for reviewing this PR and for your feedback, I hope it was not that painful. I addressed your comments, could you please take a look again?

@kulyk yes, this toggle is really inconvenient, thanks for pointing. If you don't mind, I'll fix it in a separate PR, because it worked the same way before.

@daltojohnso I added a bit of tests on selectors, but previously they contained some mutating logic. Instead, I have cypress specs for that, but the coverage can be increased. Would you mind if I add more tests in a follow-up PR?

flamber reviewed (commented):
Testing `909e299`
When I try to edit a sandbox, then it show 404 and changes to URL to this:
`/undefined/segmented/group/[object Object]`

alxnddr committed:
fix edit sandbox access action

alxnddr committed:
add cypress spec for sandboxing

alxnddr commented:
@flamber oh, seems like I broke it during some refactoring while addressing review comments. Fixed, added a spec to prevent breaking it again
